### PR TITLE
Docs: Add basic Gateway API troubleshooting

### PR DESCRIPTION
**Pull request by meyskens**
This adds a very short getting started to troubleshooting Gateway API part of the documentation

```release-note
Add a troubleshooting Gateway API part of the documentation
```

### CODE REVIEW

**Review by youngnick — APPROVED** [review]
This is a great overview, thanks!

File: Documentation/network/servicemesh/gateway-api/mistakes-warning.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+.. warning::
+
+   Gateway API is a recent addition to Kubernetes. As it is not yet widely used we have not yet had a chances to get lots of user feedback.
+   If you ran into an issue which was a mistake and not yet listed here we highly appreciate a PR to add yours to the list!
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/mistakes-warning.rst]: "We" is unclear, specify as "the Cilium project"; "mistake" is unclear; use present tense; use clear, plain language
```suggestion
   Gateway API is a recent addition to Kubernetes, and the Cilium project has not yet received much user feedback.
   If you encounter an issue that is not yet listed in this section, consider opening a PR to add yours to the list.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 2]
```diff
@@ -0,0 +1,164 @@
+These steps are for troubleshooting issues with Gateway API resources. 
+It is important to also have followed the Generic and Setup Verification steps from the :ref:`Troubleshooting Ingress & Service Mesh page <troubleshooting_servicemesh>`  
+This part of the documentation will guide you through the different mechanics of Gateway API and how to troubleshoot them.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Use clear, plain language
```suggestion
This page guides you through the different mechanics of Gateway API and how to troubleshoot them.

Be sure to follow the Generic and Setup Verification steps from the :ref:`Troubleshooting Ingress & Service Mesh page<troubleshooting_servicemesh>`.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+Checking the resources
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]:
```suggestion
Checking resources
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+#. Check the Gateway resource 
+
+    .. code-block:: shell-session
+
+      $ kubectl get gateway -A
+      NAMESPACE                   NAME                 CLASS    ADDRESS          PROGRAMMED   AGE
+      website                     http-gateway         cilium   172.21.255.202   True         5h
+      webshop                     tls-gateway          cilium   172.21.255.203   True         5h
+  
+    This will get you an overview for all the Gateways in the cluster. There are a few important tings to check:
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Specific is better than general, use present tense, typo
```suggestion
    The preceding command returns an overview of all the Gateways in the cluster. Check the following:
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+    * Is the Gateway programmed? This means that Cilium saw the gateway and prepared configuration for it.
+      
+      * If this is missing make sure that Gateway API is enabled in the Cilium configuration.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: You are overusing "this" and it isn't clear what you mean each time. For example:
> If this is missing

If what is missing, specifically?
```suggestion
    * Is the Gateway programmed? 
    
        A programmed Gateway means that Cilium saw the gateway and prepared a configuration for it.
      
      * If this is missing make sure that Gateway API is enabled in the Cilium configuration.
```

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@meyskens 👋🏻 It's clear that a lot of enthusiasm and care went into this documentation. ✨ That said, I'm returning this PR to you because it needs more work before it's in a reviewable state for documentation. 

Specifically, you must replace every instance of "this" with a clear indication of what specifically you're referring to. It's unreasonable to expect that a reviewer, let alone a user, will correctly infer what you mean by indirect reference.

**Comment by youngnick** [comment]
ping @networkop to keep an eye on this one

**Comment by meyskens** [comment]
Thanks a lot for the feedback! Improved it, hope this is better :)

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      website                     http-gateway         cilium   172.21.255.202   True         5h
+      webshop                     tls-gateway          cilium   172.21.255.203   True         5h
+  
+    The preceding command returns an overview of all the Gateways in the cluster. Check the following:
+
+    * Is the Gateway programmed? 
+    
+        A programmed Gateway means that Cilium saw the gateway and prepared a configuration for it.
+      
+      * If the ``Programmed true`` indicator is missing make sure that Gateway API is enabled in the Cilium configuration.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: comma nit
```suggestion
      * If the ``Programmed true`` indicator is missing, make sure that Gateway API is enabled in the Cilium configuration.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  
+    * Does the gateway have an address? If you see one it means that that a LoadBalancer service got assigned to the gateway.
+  
+      * You can check the service with ``kubectl get service``, if there is no IP there either you might be missing a LoadBalancer implementation.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Formatting, clarity, present tense
```suggestion
    * Does the gateway have an address?
    
    You can check the service with ``kubectl get service``. 
    If the gateway has an address, it means that a LoadBalancer service is assigned to the gateway.
    If no IP appears, you might be missing a LoadBalancer implementation.
  

```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+      $ kubectl get gateway -A
+      NAMESPACE                   NAME                 CLASS    ADDRESS          PROGRAMMED   AGE
+      website                     http-gateway         cilium   172.21.255.202   True         5h
+      webshop                     tls-gateway          cilium   172.21.255.203   True         5h
+  
+    The preceding command returns an overview of all the Gateways in the cluster. Check the following:
+
+    * Is the Gateway programmed? 
+    
+        A programmed Gateway means that Cilium saw the gateway and prepared a configuration for it.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: clarity
```suggestion
        A programmed Gateway means that Cilium prepared a configuration for it.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+      
+      * If the ``Programmed true`` indicator is missing make sure that Gateway API is enabled in the Cilium configuration.
+  
+    * Does the gateway have an address? If you see one it means that that a LoadBalancer service got assigned to the gateway.
+  
+      * You can check the service with ``kubectl get service``, if there is no IP there either you might be missing a LoadBalancer implementation.
+  
+    * Is the class ``cilium``? Cilium will only program Gateways with the class ``cilium``.
+    * In case the resource type is not found, make sure that the Gateway API CRDs are installed.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Formatting and clarity, present tense

Resource type for what? Be specific.
```suggestion
    * Is the class ``cilium``? 
    
    Cilium only program Gateways with the class ``cilium``.
    
    * If the resource type is not found, make sure that the Gateway API CRDs are installed.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+    If you need more info to further investigate the issue you can always use ``kubectl describe gateway``
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Clarity
```suggestion
    You can use ``kubectl describe gateway`` to investigate issues more thoroughly.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            Status:                True
+            Type:                  Programmed
+            Message:               Listener Accepted
+            Reason:                Accepted
+            Status:                True
+            [...]
+
+    You'll see under the status the general status of the Gateway as well as the status of the configured listeners.
+    Per listener you also get a count of how many routes are successfully attached to it.
+    Per Gateway or listener you can always see two status conditions:
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Clear, plain language, present tense
```suggestion
    You can see the general status of the gateway as well as the status of the configured listeners.
    
    Listener status displays the number of routes successfully attached to the listener.
    
    You can see status conditions for both gateway and listener:
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+      * ``Accepted``: this means that the Gateway configuration was accepted.
+      * ``Programmed``: this means that the Gateway configuration was programmed into Envoy.
+      * ``ResolvedRefs``: this means that all referenced secrets were found and got permission to be used.
+  
+    If any of these are set to false you can get more information by looking at the ``Message`` and ``Reason`` fields.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Clarity
```suggestion
      * ``Accepted``: the Gateway configuration was accepted.
      * ``Programmed``: the Gateway configuration was programmed into Envoy.
      * ``ResolvedRefs``: all referenced secrets were found and have permission for use.
  
    If any of these conditions are set to false, the ``Message`` and ``Reason`` fields give more information.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+
+#. Check the HTTPRoute resource
+
+  When our Gateway is functional we can check our routes to see if they are configured correctly.
+  The way to do check the route status is very similar to the Gateway resource. While these instructions are written for HTTPRoute, they also apply to other route types.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Address the user as "you"
```suggestion
  When the Gateway is functional, you can check the routes to verify if they are configured correctly.
  The way to check route status is similar to checking the status of a gateway resource. 
  
  While these instructions are written for HTTPRoute, they also apply to other route types.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+  .. code-block:: shell-session
+
+    $ kubectl get httproute -A
+    NAMESPACE                 NAME              HOSTNAMES         AGE
+    website                   homepage          www.example.org   17m
+    webshop                   catalog-service                     17m
+    webshop                   cart-service                        17m
+
+  To get further information we can again use ``kubectl describe httproute <name>``
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Address the user as "you", period nit
```suggestion
  To get more information, enter ``kubectl describe httproute <name>``.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+          Type:                  ResolvedRefs
+        Controller Name:         io.cilium/gateway-controller
+        Parent Ref:
+          Group:  gateway.networking.k8s.io
+          Kind:   Gateway
+          Name:   same-namespace
+
+  Here under status we get all the conditions that are relevant for the specific HTTPRoute.
+  These are listed per parent reference to the gateway. If you linked the route to multiple gateways you will see multiple entries here.
+  There are two status conditions that are of interest to investigate:
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Clarity, address the user as "you"

There are multiple fields with "Accepted" and "ResolvedRefs". Which ones should users pay attention to?
```suggestion
  Status lists the conditions that are relevant for the specific HTTPRoute.
  Conditions are listed by parent reference to the gateway. If you linked the route to multiple gateways, multiple entries appear.
  Notice the following conditions:
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+  * ``Accepted``: This means that the HTTPRoute configuration was correct and accepted.
+  * ``ResolvedRefs``: This means that the referenced services were found and are valid references.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: clarity
```suggestion
  * ``Accepted``: The HTTPRoute configuration was correct and accepted.
  * ``ResolvedRefs``: The referenced services were found and are valid references.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+  
+  If any of these are set to false you can get more information by looking at the ``Message`` and ``Reason`` fields.
+
+Common mistakes
+---------------
+
+
+.. include:: mistakes-warning.rst
+
+* The backend service does not exist. You can check if the backend service was found under the ``conditions`` field when doing a ``kubectl describe httproute <name>``
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: formatting, period nit, active voice

What's doing the finding? Is it Kubernetes via kubectl? It's better to be as specific as possible.
```suggestion
* The backend service does not exist. 

    To verify whether the backend service was found, run ``kubectl describe httproute <name>`` and inspect the ``conditions`` field:
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+            Type:                  ResolvedRefs
+            Last Transition Time:  2023-06-06T13:55:10Z
+            Message:               Accepted HTTPRoute
+            Observed Generation:   1
+            Reason:                Accepted
+            Status:                True
+            Type:                  Accepted
+          Controller Name:         io.cilium/gateway-controller
+
+* The Gateway specified under ``parentRefs`` does not exist. You check if the gateway was found under the ``conditions`` field when doing a ``kubectl describe httproute <name>``
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Formatting nit

Similar to the previous review item ☝🏻: what's doing the finding of the gateway? Kubernetes via kubectl? It's better to be as specific as possible.
```suggestion
* The gateway specified under ``parentRefs`` does not exist.

    To verify whether the gateway was found, run ``kubectl describe httproute <name>`` and inspect the ``conditions`` field:
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+
+Underlying mechanics: a high level overview
+-------------------------------------------
+
+The deployment of Cilium has two parts that handle Gateway API resources: the Cilium agent and the Cilium operator.
+The Cilium operator will watch all Gateway API resources. After which it will verify if the resources are valid.
+If they are valid they will get marked as accepted. That starts the process of translation into Cilium Envoy Configuration resources.
+
+The Cilium Envoy Configuration resources are then picked up by the Cilium agent.
+The Cilium agent will use these to supply the built in Envoy or the Envoy DaemonSet with the configuration. Envoy is where traffic will be handled.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Clarity, present tense
```suggestion
A Cilium deployment has two parts that handle Gateway API resources: the Cilium agent and the Cilium operator.

The Cilium operator watches all Gateway API resources and verifies whether the resources are valid.
If resources are valid, the operator marks them as accepted. That starts the process of translation into Cilium Envoy Configuration resources.

The Cilium agent then picks up the Cilium Envoy Configuration resources.

The Cilium agent uses the resources to supply the configuration to the built in Envoy or the Envoy DaemonSet. Envoy handles traffic.
```

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@meyskens This is greatly improved! ✨ 

I have two questions (well, the same question in two places) about what component specifically is doing the "finding" in each troubleshooting case. PTAL and update accordingly. 

I also have a question about inspecting multiple fields with values of ``Accepted`` and ``ResourceRefs``.

The rest are formatting and clarity nits. Good work!

File: Documentation/network/servicemesh/gateway-api/mistakes-warning.rst [hunk 20]
```diff
@@ -0,0 +1,4 @@
+.. warning::
+
+   Gateway API is a recent addition to Kubernetes, and the Cilium project has not yet received much user feedback.
+   If you encounter an issue that is not yet listed in this section, consider opening a PR to add yours to the list.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/mistakes-warning.rst]: Fiddly wording nit for the sake of clarity
```suggestion
   If you encounter an issue that is not yet listed in this section, consider opening a PR to add your issue to the list.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+      
+      * If the ``Programmed true`` indicator is missing, make sure that Gateway API is enabled in the Cilium configuration.
+  
+    * Does the gateway have an address?
+    
+    You can check the service with ``kubectl get service``. 
+    If the gateway has an address, it means that a LoadBalancer service is assigned to the gateway.
+    If no IP appears, you might be missing a LoadBalancer implementation.
+  
+
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Remove new lines

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+  
+    * Is the class ``cilium``? 
+    
+    Cilium only program Gateways with the class ``cilium``.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: s/program/programs

Comment (zacharysarah) [reply Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: I've opened a separate issue to talk about capitalization of API objects, in #26904.

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+    
+    You can see status conditions for both gateway and listener:
+
+      * ``Accepted``: the Gateway configuration was accepted.
+      * ``Programmed``: the Gateway configuration was programmed into Envoy.
+      * ``ResolvedRefs``: all referenced secrets were found and have permission for use.
+  
+    If any of these conditions are set to false, the ``Message`` and ``Reason`` fields give more information.
+
+
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Remove newline
```suggestion
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+        Controller Name:         io.cilium/gateway-controller
+        Parent Ref:
+          Group:  gateway.networking.k8s.io
+          Kind:   Gateway
+          Name:   same-namespace
+
+  Status lists the conditions that are relevant for the specific ``HTTPRoute``.
+  Conditions are listed by parent reference to the gateway. If you linked the route to multiple gateways, multiple entries appear.
+  You will see a group of ``Reason``, ``Type``, ``Status`` and ``Message``. Where ``Type`` will indicate the condition type and ``Status``
+  will tell you it is met in a boolean. Optionally, ``Message`` will give you more information about the condition.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: Present tense, new line
```suggestion
  Conditions include ``Reason``, ``Type``, ``Status`` and ``Message``. ``Type`` indicates the condition type, and ``Status`` indicates with a boolean whether the condition type is met. Optionally, ``Message`` gives you more information about the condition.
  
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+  Notice the following condition types:
+
+  * ``Accepted``: The HTTPRoute configuration was correct and accepted.
+  * ``ResolvedRefs``: The referenced services were found and are valid references.
+  
+  If any of these are set to false you can get more information by looking at the ``Message`` and ``Reason`` fields.
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst]: comma nit
```suggestion
  If any of these are set to false, you can get more information by looking at the ``Message`` and ``Reason`` fields.
```

File: Documentation/network/servicemesh/gateway-api/troubleshooting.rst [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+
+Common mistakes
+---------------
+
```
Comment (zacharysarah) [inline Documentation/network/servicemesh/gateway-api/troubleshooting.rst:131]: Remove newline
```suggestion
```

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@meyskens Some final nits, then LGTM

**Review by zacharysarah — APPROVED** [review]
Thank you @meyskens! 🎉

**Pull request merged**